Support for playing impact effects

KernalsEgg opened the pull request:
I have added and expanded upon various classes and structs to enable the rewriting of the [PlayImpactEffect](https://www.creationkit.com/index.php?title=PlayImpactEffect_-_ObjectReference) function at 0x1405A2C60. They have been thoroughly tested in a plugin that I am working on. I will try to link to where I am using each of these additions for added context. Effort has been made to fit the existing style of CommonLibSSE.

**BGSDecalGroup.h**
The default values for each of the members of the DECAL_CREATION_DATA struct have been taken from 0x1401AFCD0, which has been called to initialise the struct at locations where I have seen it used.
[DECAL_CREATION_DATA](https://github.com/KernalsEgg/Trails/blob/main/src/BGSImpactManager.cpp#L228)

**BGSDecalManager.h**
BGSDecalManager::IsApplyingDecal and BGSDecalManager::SetApplyingDecal are used to set a byte that is then read by [bhkWorld::CastRay()](https://github.com/KernalsEgg/CommonLibSSE/blob/313bb7949fb4029f3cfe08902042627d740d65ee/include/RE/B/bhkWorld.h#L19) at 0x140DA75B5. ApplyDecal has been tested, but I have rewritten it so that I can ignore the [fDecalLOD0 INI setting](https://stepmodifications.org/wiki/Guide:Skyrim_INI/Display#fDecalLOD0) without forcing the decal to be applied.
[BGSDecalManager](https://github.com/KernalsEgg/Trails/blob/main/src/BGSDecalManager.cpp#L31)
[BGSDecalManager::GetSingleton()](https://github.com/KernalsEgg/Trails/blob/main/src/BGSDecalManager.cpp#L7)
[BGSDecalManager::IsApplyingDecal()](https://github.com/KernalsEgg/Trails/blob/main/src/BGSImpactManager.cpp#L37)
[BGSDecalManager::SetApplyingDecal()](https://github.com/KernalsEgg/Trails/blob/main/src/BGSImpactManager.cpp#L38) ([2](https://github.com/KernalsEgg/Trails/blob/main/src/BGSImpactManager.cpp#L303))
[BGSDecalManager::ShouldApplyDecal()](https://github.com/KernalsEgg/Trails/blob/main/src/BGSImpactManager.cpp#L226)
[BGSDecalManager::ApplyDecal()](https://github.com/KernalsEgg/Trails/blob/main/src/BGSDecalManager.cpp#L5)

**BGSImpactDataSet.h**
[BGSImpactDataSet::GetImpactData()](https://github.com/KernalsEgg/Trails/blob/main/src/BGSImpactManager.cpp#L202)

**BGSImpactManager.h**
I have similarly rewritten, but tested BGSImpactManager::PlayImpactEffect().
[IMPACT_SOUND_DATA](https://github.com/KernalsEgg/Trails/blob/main/src/BGSImpactManager.cpp#L211)
[BGSImpactManager::GetSingleton()](https://github.com/KernalsEgg/Trails/blob/main/src/BGSImpactManager.cpp#L218)
[BGSImpactManager::PlayImpactEffect()](https://github.com/KernalsEgg/Trails/blob/main/src/BGSImpactManager.cpp#L7)
[BGSImpactManager::PlaySound()](https://github.com/KernalsEgg/Trails/blob/main/src/BGSImpactManager.cpp#L218)

**BGSMaterialType.h**
[BGSMaterialType::GetMaterialType()](https://github.com/KernalsEgg/Trails/blob/main/src/BGSImpactManager.cpp#L198)

**bhkCharacterController.h**
[bhkCharacterController::GetCollisionFilterInfo()](https://github.com/KernalsEgg/Trails/blob/main/src/BGSImpactManager.cpp#L94)

**bhkCollisionObject.h**
[bhkCollisionObject::GetRigidBody()](https://github.com/KernalsEgg/Trails/blob/main/src/BGSImpactManager.cpp#L107)

**bhkShape.h**
[bhkShape::GetMaterialID()](https://github.com/KernalsEgg/Trails/blob/main/src/BGSImpactManager.cpp#L189)

**bhkWorld.h**
You multiply [in-game Units](https://www.creationkit.com/index.php?title=Unit) by bhkWorld::GetWorldScale() for use in Havok related functions. While I don't have an example of it being used, the opposite is true for bhkWorld::GetWorldScaleInverse().
[bhkWorld::CastRay()](https://github.com/KernalsEgg/Trails/blob/main/src/BGSImpactManager.cpp#L122)
[bhkWorld::GetWorldScale()](https://github.com/KernalsEgg/Trails/blob/main/src/BGSImpactManager.cpp#L82)
bhkWorld::GetWorldScaleInverse()

**bhkWorldObject.h**
[bhkWorldObject::GetPropertyValue()](https://github.com/KernalsEgg/Trails/blob/main/src/BGSImpactManager.cpp#L151)

**hkpClosestRayHitCollector.h**
[hkpClosestRayHitCollector : hkpRayHitCollector](https://github.com/KernalsEgg/Trails/blob/main/src/BGSImpactManager.cpp#L79) ([2](https://github.com/KernalsEgg/Trails/blob/main/src/BGSImpactManager.cpp#L124)) ([3](https://github.com/KernalsEgg/Trails/blob/main/src/BGSImpactManager.cpp#L177))

**hkpCollidable.h**
If the ray cast hits a reference (rather than terrain), then hkpCollidable::Get3D() is used to get the 3D of the rootCollidable. I have rewritten this using 0x140DAD060. The lower 2 bytes of the collision filter info tells you what the collidable collides with (the collision layer, RE::COL_LAYER), and the upper 2 bytes of the collision filter info tells you what it belongs to (e.g. whether it is terrain). It's worth mentioning that RE::COL_LAYER::kUnused0 and RE::COL_LAYER::kUnused1 are used as custom picks sometimes when ray casting.
[hkpCollidable::BelongsTo::kTerrain](https://github.com/KernalsEgg/Trails/blob/main/src/BGSImpactManager.cpp#L135)
[hkpCollidable::CollisionFilterInfo](https://github.com/KernalsEgg/Trails/blob/main/src/BGSImpactManager.cpp#L135)
[hkpCollidable::Get3D()](https://github.com/KernalsEgg/Trails/blob/main/src/BGSImpactManager.cpp#L193)

**hkpWorldObject.h**
[hkpWorldObject::GetPropertyValue()](https://github.com/KernalsEgg/CommonLibSSE/blob/313bb7949fb4029f3cfe08902042627d740d65ee/include/RE/B/bhkWorldObject.cpp#L9) ([2](https://github.com/KernalsEgg/CommonLibSSE/blob/313bb7949fb4029f3cfe08902042627d740d65ee/include/RE/H/hkpCollidable.cpp#L16))

**hkpWorldRayCastInput.h**
I have expanded the hkpWorldRayCastInput struct because I think it makes sense contextually. It is used as input for the [bhkWorld::CastRay()](https://github.com/KernalsEgg/CommonLibSSE/blob/313bb7949fb4029f3cfe08902042627d740d65ee/include/RE/B/bhkWorld.h#L19) function. However, it is also possible that it instead uses another struct that inherits hkpWorldRayCastInput.
[hkpWorldRayCastInput](https://github.com/KernalsEgg/Trails/blob/main/src/BGSImpactManager.cpp#L77) ([2](https://github.com/KernalsEgg/Trails/blob/main/src/BGSImpactManager.cpp#L122))

**NiAVObject.h**
[NiAVObject::GetCollisionObject()](https://github.com/KernalsEgg/Trails/blob/main/src/BGSImpactManager.cpp#L103)

**NiMatrix3.h**
Given the layout of entries in the NiMatrix3 class, it made sense to add the operator there, rather than the NiPoint3 class. I have rewritten 0x140C55AE0 more concisely as NiMatrix3::EulerAnglesToAxesZXY().
[NiMatrix3::operator*](https://github.com/KernalsEgg/Trails/blob/main/src/BGSImpactManager.cpp#L56) ([2](https://github.com/KernalsEgg/Trails/blob/main/src/BGSImpactManager.cpp#L66))
[NiMatrix3::EulerAnglesToAxesZXY()](https://github.com/KernalsEgg/Trails/blob/main/src/BGSImpactManager.cpp#L64) ([2](https://github.com/KernalsEgg/Trails/blob/main/src/BGSImpactManager.cpp#L259))

**NiNode.h**
[DECAL_APPLICATION_DATA](https://github.com/KernalsEgg/Trails/blob/main/src/BGSDecalManager.cpp#L26)
[NiNode::ApplyDecal()](https://github.com/KernalsEgg/Trails/blob/main/src/BGSDecalManager.cpp#L38)

**TES.h**
[TES::GetCell()](https://github.com/KernalsEgg/Trails/blob/main/src/BGSImpactManager.cpp#L264)
[TES::GetMaterialID()](https://github.com/KernalsEgg/Trails/blob/main/src/BGSImpactManager.cpp#L139)

**TESObjectCELL.h**
The [Impact NIF](https://www.creationkit.com/index.php?title=ImpactData) from the ImpactData form is placed using TESObjectCELL::PlaceTempEffectParticle(). The second argument appears to be unused as you can see at 0x14044B9A0, and so I have named it as such, and give it a default type.
[TESObjectCELL::PlaceTempEffectParticle()](https://github.com/KernalsEgg/Trails/blob/main/src/BGSImpactManager.cpp#L209)

Ryan-rsm-McKenzie reviewed (changes requested):
Please don't add offsets to the offset header. It is a relic from before dynamic id relocation and only serves to bloat compile time.

Ryan-rsm-McKenzie reviewed (changes requested):
Please do not add source files unless necessary. CommonLib currently has several hundred trivial source files that severely bloat compile time. Please prefer to inline the definition in the header when possible.

KernalsEgg commented:
I have removed each of my offsets from the offset header and moved each of my functions to their respective header files so long as they did not require any includes. Please let me know if I have made any other silly decisions because while I am doing my best, I am new to C++.
